### PR TITLE
feat: match the data channel IP to the control channel IP

### PR DESCRIPTION
**Pull request by Nonius**
To prevent malicious transfers the data channel IP must be the same as
 the control channel IP. If they don't match up. The data channel will
 be closed and handling will be stopped.

### CODE REVIEW

File: src/server/datachan.rs [hunk 1]
```diff
@@ -267,6 +267,29 @@ where
 {
     // We introduce a block scope here to keep the lock on the session minimal. We basically copy the needed info
     // out and then unlock.
+
+    match socket.peer_addr() {
```
Comment (hannesdejager) [inline src/server/datachan.rs:271]: OK for now, but in this method we are waiting twice now for the session lock, we could move this into a method and call it from the below block.

File: src/server/datachan.rs [hunk 2]
```diff
@@ -267,6 +267,29 @@ where
...
+        Ok(datachan_addr) => {
+            let controlcahn_ip = session_arc.lock().await.source.ip();
+            if controlcahn_ip != datachan_addr.ip() {
+                if let Err(err) = socket.shutdown(std::net::Shutdown::Both) {
```
Comment (hannesdejager) [inline src/server/datachan.rs:275]: Also log when there is no shutdown error.

File: src/server/proxy_protocol.rs [hunk 3]
```diff
@@ -35,20 +36,13 @@ pub enum ProxyError {
...
 
 impl ConnectionTuple {
-    fn new(from_ip: IpAddr, from_port: u16, to_ip: IpAddr, to_port: u16) -> Self {
-        ConnectionTuple {
-            from_ip,
-            from_port,
-            to_ip,
-            to_port,
-        }
+    pub fn hash(&self) -> String {
```
Comment (hannesdejager) [inline src/server/proxy_protocol.rs:44]: A side comment for another day: This is hardly a `hash`.  Suggest using the name `key` rather as method name.

Comment (Nonius) [reply src/server/proxy_protocol.rs]: renamed

File: src/server/datachan.rs (hunk 2 above)
Comment (Nonius) [reply src/server/datachan.rs:275]: added

**Review by hannesdejager — APPROVED** [review]

File: src/server/datachan.rs [hunk 4]
```diff
@@ -268,37 +268,38 @@ where
...
-                    )
+    let command_executor = {
+        let mut session = session_arc.lock().await;
+
+        match socket.peer_addr() {
+            Ok(datachan_addr) => {
+                let controlcahn_ip = session.source.ip();
+                if controlcahn_ip != datachan_addr.ip() {
+                    if let Err(err) = socket.shutdown(std::net::Shutdown::Both) {
+                        slog::info!(
```
Comment (hannesdejager) [inline src/server/datachan.rs:279]: This should be logged as error no?

Comment (Nonius) [reply src/server/datachan.rs]: changed

File: src/server/datachan.rs (hunk 1 above)
Comment (Nonius) [reply src/server/datachan.rs]: moved things around, please check if this is better

Comment (hannesdejager) [reply src/server/datachan.rs]: Good if it compiles :-)

**Pull request merged**